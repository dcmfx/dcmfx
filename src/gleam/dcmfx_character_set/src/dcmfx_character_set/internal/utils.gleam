import gleam/string

/// Returns the codepoint for the replacement character '�' that is emitted when
/// invalid string data is encountered.
///
pub fn replacement_character() -> UtfCodepoint {
  let assert Ok(codepoint) = string.utf_codepoint(0xFFFD)
  codepoint
}

/// Converts an integer codepoint value to a `UtfCodepoint`. The replacement
/// character '�' is returned if the integer is not a valid codepoint.
///
pub fn int_to_codepoint(i: Int) -> UtfCodepoint {
  case string.utf_codepoint(i) {
    Ok(codepoint) -> codepoint
    Error(Nil) -> replacement_character()
  }
}
