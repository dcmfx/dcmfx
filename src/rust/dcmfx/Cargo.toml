[package]
name = "dcmfx"
version = "0.17.0"
description = "Library for working with DICOM and DICOM JSON"
categories = ["wasm"]

repository.workspace = true
edition.workspace = true
authors.workspace = true
license.workspace = true
homepage.workspace = true
keywords.workspace = true

[dependencies]
dcmfx_anonymize = { path = "../dcmfx_anonymize" }
dcmfx_character_set = { path = "../dcmfx_character_set", default-features = false }
dcmfx_core = { path = "../dcmfx_core", default-features = false }
dcmfx_json = { path = "../dcmfx_json", default-features = false }
dcmfx_p10 = { path = "../dcmfx_p10", default-features = false }
dcmfx_pixel_data = { path = "../dcmfx_pixel_data", default-features = false }

[dev-dependencies]
either = "1.15.0"
image = "0.25.6"
rand = { version = "0.9.0", features = ["small_rng"] }
serde = "1.0.219"
serde_json = "1.0.140"
walkdir = "2.5.0"

[features]
default = ["std"]
std = [
  "dcmfx_character_set/std",
  "dcmfx_core/std",
  "dcmfx_json/std",
  "dcmfx_p10/std",
  "dcmfx_pixel_data/std",
]
