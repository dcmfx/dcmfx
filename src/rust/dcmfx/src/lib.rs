//! DCMfx is a collection of libraries and a CLI tool for working with the DICOM
//! standard, the international standard for medical images and related
//! information.

/// Anonymization of data sets by removing data elements that identify the
/// patient, or potentially contribute to identification of the patient.
///
/// This module is a re-export of the `dcmfx_anonymize` crate.
///
pub mod anonymize {
  pub use dcmfx_anonymize::*;
}

/// Provides core DICOM concepts including data sets, data elements, value
/// representations, transfer syntaxes, and a dictionary of the data elements
/// defined in DICOM PS3.6 as well as well-known private data elements.
///
/// This module is a re-export of the `dcmfx_core` crate.
///
pub mod core {
  pub use dcmfx_core::*;
}

/// Converts between DICOM data sets and DICOM JSON.
///
/// This module is a re-export of the `dcmfx_json` crate.
///
pub mod json {
  pub use dcmfx_json::*;
}

/// Reads and writes the DICOM Part 10 (P10) binary format used to store and
/// transmit DICOM-based medical imaging information.
///
/// This module is a re-export of the `dcmfx_p10` crate.
///
pub mod p10 {
  pub use dcmfx_p10::*;
}

/// Extracts frames of pixel data present in a data set.
///
/// This module is a re-export of the `dcmfx_pixel_data` crate.
///
pub mod pixel_data {
  pub use dcmfx_pixel_data::*;
}

mod integration_tests;
