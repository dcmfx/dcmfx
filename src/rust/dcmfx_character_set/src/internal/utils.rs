/// The `char` for the replacement character '�' that is emitted when invalid
/// string data is encountered.
///
pub(crate) const REPLACEMENT_CHARACTER: char = '�';

/// Converts an integer codepoint value to a `char`. The replacement character
/// '�' is returned if the integer is not a valid codepoint.
///
pub(crate) fn codepoint_to_char(codepoint: u32) -> char {
  char::from_u32(codepoint).unwrap_or(REPLACEMENT_CHARACTER)
}
