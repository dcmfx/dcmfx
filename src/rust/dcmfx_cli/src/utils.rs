use std::{
  fs::File,
  io::Write,
  path::{Path, PathBuf},
  sync::atomic::{AtomicBool, Ordering},
};

use dcmfx::p10::P10Error;

/// Opens an output stream for the given path, first checking whether it exists
/// and prompting the user about overwriting it if necessary. This prompt isn't
/// presented to the user if `overwrite` is true.
///
/// The path "-" is interpreted as writing to stdout.
///
pub fn open_output_stream(
  path: &PathBuf,
  display_path: Option<&PathBuf>,
  overwrite: bool,
) -> Result<Box<dyn std::io::Write>, P10Error> {
  if *path == PathBuf::from("-") {
    Ok(Box::new(std::io::stdout()))
  } else {
    if let Some(display_path) = display_path {
      println!("Writing \"{}\" â€¦", display_path.display());
    }

    if !overwrite {
      prompt_to_overwrite_if_exists(path);
    }

    match File::create(path) {
      Ok(file) => Ok(Box::new(file)),

      Err(e) => Err(P10Error::FileError {
        when: "Opening file".to_string(),
        details: e.to_string(),
      }),
    }
  }
}

/// Stores whether the user has requested to overwrite all files instead of
/// prompting for each one.
static OVERWRITE_ALL_FILES: AtomicBool = AtomicBool::new(false);

/// Prompts the user about overwriting the given file if it exists.
///
pub fn prompt_to_overwrite_if_exists(path: &Path) {
  if !path.exists() {
    return;
  }

  if OVERWRITE_ALL_FILES.load(Ordering::Relaxed) {
    return;
  }

  print!(
    "File \"{}\" already exists. Overwrite? ([y]es, [n]o, [a]ll): ",
    path.display()
  );
  std::io::stdout().flush().unwrap();

  let mut input = String::new();
  std::io::stdin().read_line(&mut input).unwrap();
  let input = input.trim().to_lowercase();

  if input != "y" && input != "yes" && input != "a" && input != "all" {
    std::process::exit(1)
  }

  if input == "a" || input == "all" {
    OVERWRITE_ALL_FILES.store(true, Ordering::Relaxed);
  }
}

/// Appends a suffix to a path.
///
pub fn path_append(mut path: PathBuf, suffix: &str) -> PathBuf {
  path.set_file_name(format!(
    "{}{}",
    path
      .file_name()
      .unwrap_or(std::ffi::OsStr::new(""))
      .to_string_lossy(),
    suffix
  ));

  path
}
