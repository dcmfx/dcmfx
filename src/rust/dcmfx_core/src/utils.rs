#[cfg(feature = "std")]
use std::rc::Rc;

#[cfg(not(feature = "std"))]
use alloc::{format, rc::Rc, string::String, vec, vec::Vec};

/// A reference counted byte slice holds an `Rc<Vec<u8>>` to a shareable buffer,
/// along with a range that specifies the part of that buffer that this slice
/// refers to.
///
/// This type is used widely to avoid copying buffers wherever possible, and in
/// most cases can be used like a `&[u8]` would be.
///
#[derive(Clone)]
pub struct RcByteSlice {
  data: Rc<Vec<u8>>,
  range: core::ops::Range<usize>,
}

impl RcByteSlice {
  /// Creates a new referenced counted byte slice from a `Vec<u8>`.
  ///
  pub fn from_vec(data: Vec<u8>) -> Self {
    let range = 0..data.len();

    Self {
      data: Rc::new(data),
      range,
    }
  }

  /// Creates an empty reference counted byte slice.
  ///
  pub fn empty() -> Self {
    Self {
      data: Rc::new(vec![]),
      range: 0..0,
    }
  }

  /// Slices this referenced counted byte slice, returning a new reference
  /// counted byte slice that points to the same underlying data.
  ///
  /// This function does not copy any data.
  ///
  pub fn slice(&self, start: usize, end: usize) -> Self {
    assert!(start <= end, "Byte slice range out of bounds");
    assert!(end <= self.range.len(), "Byte slice range out of bounds");

    Self {
      data: self.data.clone(),
      range: (self.range.start + start)..(self.range.start + end),
    }
  }

  /// Returns a new reference counted byte slice with the specified number of
  /// bytes dropped from the front.
  ///
  pub fn drop(&self, n: usize) -> Self {
    self.slice(n, self.len())
  }

  /// Returns a new reference counted byte slice to the specified number of
  /// leading bytes.
  ///
  pub fn take(&self, n: usize) -> Self {
    self.slice(0, n)
  }

  /// Consumes this reference counted byte slice and turns it into a `Vec<u8>`.
  /// Avoids a copy when possible.
  ///
  /// This function copies data if there are multiple references to the
  /// underlying buffer, or its slice bounds do not cover the whole buffer.
  ///
  pub fn into_vec(self) -> Vec<u8> {
    if self.range == (0..self.data.len()) {
      match Rc::try_unwrap(self.data) {
        Ok(data) => data,
        Err(data_rc) => data_rc[self.range.clone()].to_vec(),
      }
    } else {
      self.as_slice().to_vec()
    }
  }

  fn as_slice(&self) -> &[u8] {
    &self.data[self.range.clone()]
  }
}

impl core::fmt::Debug for RcByteSlice {
  fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
    write!(f, "{:?}", self.as_slice())
  }
}

impl PartialEq for RcByteSlice {
  fn eq(&self, other: &Self) -> bool {
    self.as_slice() == other.as_slice()
  }
}

impl core::ops::Deref for RcByteSlice {
  type Target = [u8];

  fn deref(&self) -> &Self::Target {
    self.as_slice()
  }
}

impl From<Vec<u8>> for RcByteSlice {
  fn from(vec: Vec<u8>) -> Self {
    RcByteSlice::from_vec(vec)
  }
}

/// Inspects a byte slice in hexadecimal, e.g. `[1A 2B 3C 4D]`. If the number of
/// bytes in the slice exceeds `max_length` then not all bytes will be
/// shown and a trailing ellipsis will be appended, e.g. `[1A 2B 3C 4D …]`.
///
pub fn inspect_u8_slice(bytes: &[u8], max_length: usize) -> String {
  let byte_count = core::cmp::min(max_length, bytes.len());

  let s = bytes[0..byte_count]
    .iter()
    .map(|byte| format!("{:02X}", byte))
    .collect::<Vec<_>>()
    .join(" ");

  if byte_count == bytes.len() {
    format!("[{}]", s)
  } else {
    format!("[{} …]", s)
  }
}

#[cfg(test)]
mod tests {
  use super::*;

  #[cfg(not(feature = "std"))]
  use alloc::string::ToString;

  #[test]
  fn inspect_u8_slice_test() {
    assert_eq!(
      inspect_u8_slice(&[0xD1, 0x96, 0x33], 100),
      "[D1 96 33]".to_string()
    );

    assert_eq!(
      inspect_u8_slice(&[0xD1, 0x96, 0x33, 0x44], 3),
      "[D1 96 33 …]".to_string()
    );
  }
}
