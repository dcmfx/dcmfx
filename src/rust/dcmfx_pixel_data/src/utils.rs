use core::ops::{Add, Div};

/// Unsigned integer division that rounds to the nearest integer instead of
/// truncating towards zero like standard integer division.
///
#[inline]
pub fn udiv_round<T>(numer: T, denom: T) -> T
where
  T: Copy + Add<Output = T> + Div<Output = T> + From<u8>,
{
  (numer + (denom / T::from(2))) / denom
}
