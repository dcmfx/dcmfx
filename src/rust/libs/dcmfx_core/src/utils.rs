/// Removes all whitespace from the end of the passed string. Whitespace is
/// defined as the following Unicode codepoints: U+0000, U+0009, U+000A, U+000D,
/// U+0020.
///
pub fn trim_right_whitespace(s: &str) -> &str {
  s.trim_end_matches([
    '\u{0000}', '\u{0009}', '\u{000A}', '\u{000D}', '\u{0020}',
  ])
}
