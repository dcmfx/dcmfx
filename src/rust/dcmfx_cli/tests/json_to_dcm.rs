mod utils;

use assert_cmd::Command;
use dcmfx::{
  core::DataSet,
  json::{DataSetJsonExtensions, DicomJsonConfig},
  p10::DataSetP10Extensions,
};
use insta::assert_snapshot;
use utils::{generate_temp_filename, get_stdout};

const JSON_CONFIG: DicomJsonConfig = DicomJsonConfig {
  store_encapsulated_pixel_data: true,
  pretty_print: true,
};

#[test]
fn with_output_filename() {
  let dicom_json_file =
    "../../../test/assets/pydicom/test_files/SC_rgb_small_odd.dcm.json";
  let temp_path = generate_temp_filename();

  let mut cmd = Command::cargo_bin("dcmfx_cli").unwrap();
  cmd
    .arg("json-to-dcm")
    .arg(dicom_json_file)
    .arg("--output-filename")
    .arg(&temp_path)
    .arg("--implementation-version-name")
    .arg("DCMfx Test")
    .assert()
    .success()
    .stdout(format!("Writing \"{}\" …\n", temp_path.display()));

  let mut cmd = Command::cargo_bin("dcmfx_cli").unwrap();
  let assert = cmd.arg("print").arg(temp_path).assert();

  assert_snapshot!("with_output_filename", get_stdout(assert));
}

#[test]
fn with_stdout_output() {
  let dicom_json_file =
    "../../../test/assets/pydicom/test_files/SC_rgb_small_odd.dcm.json";

  let mut cmd = Command::cargo_bin("dcmfx_cli").unwrap();
  let assert = cmd
    .arg("json-to-dcm")
    .arg(dicom_json_file)
    .arg("--output-filename")
    .arg("-")
    .assert()
    .success();

  let p10_data = assert.get_output().stdout.clone();

  // Convert stdout data back to JSON so it can be asserted
  let json = DataSet::read_p10_bytes(p10_data.into())
    .unwrap()
    .to_json(JSON_CONFIG)
    .unwrap();

  assert_snapshot!("with_stdout_output", json);
}

#[test]
fn with_output_directory() {
  let dicom_json_files = [
    "../../../test/assets/pydicom/test_files/SC_rgb_small_odd.dcm.json",
    "../../../test/assets/pydicom/test_files/SC_rgb_small_odd_jpeg.dcm.json",
  ];

  let output_directory = generate_temp_filename();
  std::fs::create_dir(&output_directory).unwrap();
  let output_files = [
    output_directory.join("SC_rgb_small_odd.dcm.json.dcm"),
    output_directory.join("SC_rgb_small_odd_jpeg.dcm.json.dcm"),
  ];

  let mut cmd = Command::cargo_bin("dcmfx_cli").unwrap();
  cmd
    .arg("json-to-dcm")
    .args(dicom_json_files)
    .arg("--output-directory")
    .arg(output_directory)
    .assert()
    .success()
    .stdout(format!(
      "Writing \"{}\" …\nWriting \"{}\" …\n",
      output_files[0].display(),
      output_files[1].display()
    ));

  assert_snapshot!(
    "with_output_directory_0",
    DataSet::read_p10_file(&output_files[0])
      .unwrap()
      .to_json(JSON_CONFIG)
      .unwrap()
  );
  assert_snapshot!(
    "with_output_directory_1",
    DataSet::read_p10_file(&output_files[1])
      .unwrap()
      .to_json(JSON_CONFIG)
      .unwrap()
  );
}

#[test]
fn with_multiple_inputs() {
  let dicom_files = [generate_temp_filename(), generate_temp_filename()];
  let output_files = [
    format!("{}{}", dicom_files[0].to_string_lossy(), ".json"),
    format!("{}{}", dicom_files[1].to_string_lossy(), ".json"),
  ];

  std::fs::copy(
    "../../../test/assets/pydicom/test_files/SC_rgb_small_odd.dcm",
    &dicom_files[0],
  )
  .unwrap();
  std::fs::copy(
    "../../../test/assets/pydicom/test_files/SC_rgb_small_odd_jpeg.dcm",
    &dicom_files[1],
  )
  .unwrap();

  let mut cmd = Command::cargo_bin("dcmfx_cli").unwrap();
  cmd
    .arg("dcm-to-json")
    .args(dicom_files)
    .arg("--pretty")
    .assert()
    .success()
    .stdout(format!(
      "Writing \"{}\" …\nWriting \"{}\" …\n",
      output_files[0], output_files[1]
    ));

  assert_snapshot!(
    "with_multiple_inputs_0",
    std::fs::read_to_string(&output_files[0]).unwrap()
  );
  assert_snapshot!(
    "with_multiple_inputs_1",
    std::fs::read_to_string(&output_files[1]).unwrap()
  );
}
