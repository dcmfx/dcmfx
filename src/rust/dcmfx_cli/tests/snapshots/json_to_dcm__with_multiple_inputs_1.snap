---
source: dcmfx_cli/tests/json_to_dcm.rs
expression: "std::fs::read_to_string(&output_files[1]).unwrap()"
snapshot_kind: text
---
{
  "00020010": {
    "vr": "UI",
    "Value": [
      "1.2.840.10008.1.2.4.50"
    ]
  },
  "00080008": {
    "vr": "CS",
    "Value": [
      "DERIVED",
      "SECONDARY",
      "OTHER"
    ]
  },
  "00080016": {
    "vr": "UI",
    "Value": [
      "1.2.840.10008.5.1.4.1.1.7"
    ]
  },
  "00080018": {
    "vr": "UI",
    "Value": [
      "1.2.276.0.7230010.3.1.4.8323329.1100.1521494053.974393"
    ]
  },
  "00080020": {
    "vr": "DA",
    "Value": [
      "20170101"
    ]
  },
  "00080023": {
    "vr": "DA"
  },
  "0008002A": {
    "vr": "DT"
  },
  "00080030": {
    "vr": "TM",
    "Value": [
      "120000"
    ]
  },
  "00080033": {
    "vr": "TM"
  },
  "00080050": {
    "vr": "SH"
  },
  "00080060": {
    "vr": "CS",
    "Value": [
      "OT"
    ]
  },
  "00080064": {
    "vr": "CS",
    "Value": [
      "SYN"
    ]
  },
  "00080090": {
    "vr": "PN",
    "Value": [
      {
        "Alphabetic": "Moriarty^James"
      }
    ]
  },
  "00082111": {
    "vr": "ST",
    "Value": [
      "Lossy compression with JPEG baseline, IJG quality factor 90, compression ratio 0.084906 [Scaled image]"
    ]
  },
  "00082112": {
  "vr": "SQ",
    "Value": [
      {
        "00081150": {
          "vr": "UI",
          "Value": [
            "1.2.840.10008.5.1.4.1.1.7"
          ]
        },
        "00081155": {
          "vr": "UI",
          "Value": [
            "1.2.276.0.7230010.3.1.4.8323329.1099.1521494048.423534"
          ]
        },
        "0040A170": {
        "vr": "SQ",
          "Value": [
            {
              "00080100": {
                "vr": "SH",
                "Value": [
                  "121320"
                ]
              },
              "00080102": {
                "vr": "SH",
                "Value": [
                  "DCM"
                ]
              },
              "00080104": {
                "vr": "LO",
                "Value": [
                  "Uncompressed predecessor"
                ]
              }
            }
          ]
        }
      }
    ]
  },
  "00089215": {
  "vr": "SQ",
    "Value": [
      {
        "00080100": {
          "vr": "SH",
          "Value": [
            "113040"
          ]
        },
        "00080102": {
          "vr": "SH",
          "Value": [
            "DCM"
          ]
        },
        "00080104": {
          "vr": "LO",
          "Value": [
            "Lossy Compression"
          ]
        }
      }
    ]
  },
  "00100010": {
    "vr": "PN",
    "Value": [
      {
        "Alphabetic": "Lestrade^G"
      }
    ]
  },
  "00100020": {
    "vr": "LO",
    "Value": [
      "ID1"
    ]
  },
  "00100030": {
    "vr": "DA"
  },
  "00100040": {
    "vr": "CS",
    "Value": [
      "F"
    ]
  },
  "00101010": {
    "vr": "AS",
    "Value": [
      "024Y"
    ]
  },
  "00185100": {
    "vr": "CS"
  },
  "0020000D": {
    "vr": "UI",
    "Value": [
      "1.2.826.0.1.3680043.8.498.12406831542731051035295345080039845114"
    ]
  },
  "0020000E": {
    "vr": "UI",
    "Value": [
      "1.2.826.0.1.3680043.8.498.16157229083793556332623330502397121062"
    ]
  },
  "00200010": {
    "vr": "SH",
    "Value": [
      "1"
    ]
  },
  "00200011": {
    "vr": "IS",
    "Value": [
      1
    ]
  },
  "00200013": {
    "vr": "IS",
    "Value": [
      1
    ]
  },
  "00200020": {
    "vr": "CS"
  },
  "00200060": {
    "vr": "CS"
  },
  "00204000": {
    "vr": "LT",
    "Value": [
      "Test Image with 10 rows of (255,0,0), 10 rows of (255,128,128), 10 rows of (0,255,0), 10 rows of (128,255,128), 10 rows of (0,0,255), 10 rows of (128,128,255), 10 rows of (0,0,0), 10 rows of (64,64,64), 10 rows of (192,192,192), 10 rows of (255,255,255), uncompressed"
    ]
  },
  "00280002": {
    "vr": "US",
    "Value": [
      3
    ]
  },
  "00280004": {
    "vr": "CS",
    "Value": [
      "YBR_FULL"
    ]
  },
  "00280006": {
    "vr": "US",
    "Value": [
      0
    ]
  },
  "00280008": {
    "vr": "IS",
    "Value": [
      1
    ]
  },
  "00280010": {
    "vr": "US",
    "Value": [
      3
    ]
  },
  "00280011": {
    "vr": "US",
    "Value": [
      3
    ]
  },
  "00280030": {
    "vr": "DS",
    "Value": [
      33.333333,
      33.333333
    ]
  },
  "00280100": {
    "vr": "US",
    "Value": [
      8
    ]
  },
  "00280101": {
    "vr": "US",
    "Value": [
      8
    ]
  },
  "00280102": {
    "vr": "US",
    "Value": [
      7
    ]
  },
  "00280103": {
    "vr": "US",
    "Value": [
      0
    ]
  },
  "00282110": {
    "vr": "CS",
    "Value": [
      "01"
    ]
  },
  "00282112": {
    "vr": "DS",
    "Value": [
      0.084906
    ]
  },
  "00282114": {
    "vr": "CS",
    "Value": [
      "ISO_10918_1"
    ]
  },
  "7FE00010": {
    "vr": "OB",
    "InlineBinary": "/v8A4AQAAAAAAAAA/v8A4D4BAAD/2P/gABBKRklGAAEBAAABAAEAAP/bAEMAAwICAwICAwMDAwQDAwQFCAUFBAQFCgcHBggMCgwMCwoLCw0OEhANDhEOCwsQFhARExQVFRUMDxcYFhQYEhQVFP/bAEMBAwQEBQQFCQUFCRQNCw0UFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFP/AABEIAAMAAwMBEQACEQEDEQH/xAAUAAEAAAAAAAAAAAAAAAAAAAAG/8QAGhAAAQUBAAAAAAAAAAAAAAAABAAFFlKV0v/EABQBAQAAAAAAAAAAAAAAAAAAAAD/xAAfEQAAAwkAAAAAAAAAAAAAAAAAAgMEERJSVJKT0dL/2gAMAwEAAhEDEQA/AGcJaqmaBHaQM9OliS4B55zXG2P/2QA="
  }
}
